source/common/http/http2/codec_impl.cc

```diff
@@ -1131,14 +1135,18 @@ ConnectionImpl::ClientHttp2Options::ClientHttp2Options(
 ClientConnectionImpl::ClientConnectionImpl(
     Network::Connection& connection, Http::ConnectionCallbacks& callbacks, Stats::Scope& stats,
     const envoy::config::core::v3::Http2ProtocolOptions& http2_options,
-    const uint32_t max_response_headers_kb, const uint32_t max_response_headers_count)
+    const uint32_t max_response_headers_kb, const uint32_t max_response_headers_count,
+    bool test_only_session)
     : ConnectionImpl(connection, stats, http2_options, max_response_headers_kb,
                      max_response_headers_count),
       callbacks_(callbacks) {
   ClientHttp2Options client_http2_options(http2_options);
-  nghttp2_session_client_new2(&session_, http2_callbacks_.callbacks(), base(),
-                              client_http2_options.options());
-  sendSettings(http2_options, true);
+  // Check if a test is going to be creating its own session.
+  if (!test_only_session) {
+    nghttp2_session_client_new2(&session_, http2_callbacks_.callbacks(), base(),
```

Comment:
I think I'd mildly prefer wrapping this in a pure virtual function which is overridden in the tests, as it keeps the APIs free of test only values.  WDYT?

Reply:
+1. Also, obviously we won't be able to call a virtual function from a constructor. My advice would be to create this object using a static factory pattern, and then the test can override as needed.

Reply:
SGTM re: using a static factory for this.

Reply:
Done. PTAL.